numpy/core/tests/test_numeric.py

```diff
@@ -1775,11 +1798,94 @@ def test_clip_func_takes_out(self):
 
     def test_clip_nan(self):
         d = np.arange(7.)
-        assert_equal(d.clip(min=np.nan), d)
-        assert_equal(d.clip(max=np.nan), d)
-        assert_equal(d.clip(min=np.nan, max=np.nan), d)
-        assert_equal(d.clip(min=-2, max=np.nan), d)
-        assert_equal(d.clip(min=np.nan, max=10), d)
+        with assert_warns(DeprecationWarning):
+            assert_equal(d.clip(min=np.nan), d)
+        with assert_warns(DeprecationWarning):
+            assert_equal(d.clip(max=np.nan), d)
+        with assert_warns(DeprecationWarning):
+            assert_equal(d.clip(min=np.nan, max=np.nan), d)
+        with assert_warns(DeprecationWarning):
+            assert_equal(d.clip(min=-2, max=np.nan), d)
+        with assert_warns(DeprecationWarning):
+            assert_equal(d.clip(min=np.nan, max=10), d)
+
+    def test_object_clip(self):
+        a = np.arange(10, dtype=object)
+        actual = np.clip(a, 1, 5)
+        expected = np.array([1, 1, 2, 3, 4, 5, 5, 5, 5, 5])
+        assert actual.tolist() == expected.tolist()
+
+    def test_clip_all_none(self):
+        a = np.arange(10, dtype=object)
+        with assert_raises_regex(ValueError, 'max or min'):
+            np.clip(a, None, None)
+
+    def test_clip_invalid_casting(self):
+        a = np.arange(10, dtype=object)
+        with assert_raises_regex(ValueError,
+                                 'casting must be one of'):
+            self.fastclip(a, 1, 8, casting="garbage")
+
+    @pytest.mark.parametrize("amin, amax", [
+        # two scalars
+        (1, 0),
+        # mix scalar and array
+        (1, np.zeros(10)),
+        # two arrays
+        (np.ones(10), np.zeros(10)),
+        ])
+    def test_clip_value_min_max_flip(self, amin, amax):
+        a = np.arange(10, dtype=np.int64)
+        # requirement from ufunc_docstrings.py
+        expected = np.maximum(amin, np.minimum(a, amax))
+        actual = np.clip(a, amin, amax)
+        assert_equal(actual, expected)
+
+    @pytest.mark.parametrize("arr, amin, amax, exp", [
+        # for a bug in npy_ObjectClip, based on a
+        # case produced by hypothesis
+        (np.zeros(10, dtype=np.int64),
+         0,
+         -9223372036854775809,
+         np.zeros(10, dtype=object)),
+        # for bugs in NPY_TIMEDELTA_MAX, based on a case
+        # produced by hypothesis
+        (np.zeros(10, dtype='m8') - 1,
+         0,
+         0,
+         np.zeros(10, dtype='m8')),
+    ])
+    def test_clip_problem_cases(self, arr, amin, amax, exp):
+        actual = np.clip(arr, amin, amax)
+        assert_equal(actual, exp)
+
+    @pytest.mark.xfail(reason="no scalar nan propagation yet")
```

Comment:
need to decide if scalar `np.nan` should propagate so marking this test with xfail too for now; the function spec does propagate the `nan` here.

Historical behavior is yet another dimension I haven't fully explored, though the deprecation message implies propagation of nan has been the norm?

Reply:
Historically Nan scalars propagate only in non-native memory orders, which I've deprecated

Reply:
We talked about this briefly yesterday. And I think the consensus is, that whatever the current discussion/state is exactly, we should be be practical. Especially how incorrectly sorted min and max value for clipping behaves is OK to be undefined (or raise an error). And it is OK to risk breaking current code.
With respect to NaN, I think it does not matter much which way we go and we should just decide for one thing. Although, in the case of NaN, I guess it may make sense to do a deprecation for the scalar case.
So the gist is: Don't worry too much about some tiny inconsistencies in corner cases of corner cases, it is fine to keep some things undefined (we could even make it well defined later if necessary).

Reply:
> Although, in the case of NaN, I guess it may make sense to do a deprecation for the scalar case.

This is what I've implemented, simply because we actually had tests requiring that behavior